redirect traffic to proxy if proxy-visibility annotation is added to a pod if policy does not apply to an endpoint

ianvernon opened the pull request:
If a proxy-visibility annotation is added to an Endpoint
(`io.cilium.annotation.proxy-visibility`), and no policy applies to the endpoint
representing the pod which was annotated in the traffic direction specified
in the annotation, add a redirect to the proxy which will redirect traffic
at the given direction, L4 port / protocol, to the proxy corresponding to the
L7 parser in the annotation.

The value of the above annotation is expected to be of the following format:

```
<Traffic Direction/L4 Port/L4 Protocol/L7 Parser>,<...>
```

This format specifies a set of tuples which represent which traffic should be
redirected for the given pod which has been annotated. If policy selects the
Endpoint in the traffic direction which also has been set in the
proxy-visibility annotation, then the proxy-visibility annotation currently does
not take effect. In other words, visibility annotations only work in a given
traffic direction for a pod if no policy applies in that traffic direction.

If any of the tuples in the annotation are improperly formatted (e.g., fields
out of order, port bigger than 16 bits), an empty visibility policy is created
for the Endpoint with the annotation. This means that no L7 visibility will
be in place if any of the tuples are improperly formatted.

Update the equality check for Pods to check whether the new annotation,
`io.cilium.proxy-visibility` is present in the new set of annotations,
but not present in the old one, and vice-versa. If so, plumb the annotations
into the Endpoint itself, and trigger a regeneration.

Signed-off by: Ian Vernon <ian@cilium.io>

TODO's / follow-ups:

- [ ] Remove need to force policy computation to make sure that pointer comparison in policy map syncing fails. 
- [ ] Make L4Filter compose `VisibilityMetadata`.
- [ ] Apply visibility policy in cases where policy does apply for a given l4 port / protocol.

```release-note
Add support for L7 visibility via pod annotations
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9210)
<!-- Reviewable:end -->

joestringer commented on daemon/k8s_watcher.go:
```diff
@@ -1821,22 +1832,48 @@ func (d *Daemon) updateK8sPodV1(oldK8sPod, newK8sPod *types.Pod) error {
+		stateTransitionSucceeded := podEP.SetState(endpoint.StateWaitingToRegenerate, "annotations updated")
+		if stateTransitionSucceeded {
+			podEP.Regenerate(&regeneration.ExternalRegenerationMetadata{
+				Reason: "annotations updated",
```
Need to set the `RegenerationLevel` to `RegenerateWithoutDatapath` too.

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -278,6 +280,13 @@ type Endpoint struct {
+func (e *Endpoint) UpdateAnnotations(anno map[string]string) {
+	e.unconditionalLock()
+	defer e.unlock()
+	e.annotations = anno
```
I would likely plan to do some additional (simple) processing here on the annotations, to generate something closer to what the policy subsystem can deal with (ie `L4Policy` / `L4PolicyMap`).

joestringer reviewed (commented):
I didn't review in depth since it's still draft, but here's a couple of comments that came to mind.

aanm commented on pkg/endpoint/policy.go:
```diff
@@ -709,3 +710,47 @@ func (e *Endpoint) GetCIDRPrefixLengths() (s6, s4 []int) {
+		err error
+	)
+
+	if anno != nil {
```
no need for this check

ianvernon replied:
The function is exported / public, so any caller could provide a nil map.

aanm replied:
so? line 743 won't panic

aanm commented on pkg/endpoint/policy.go:
```diff
@@ -709,3 +710,47 @@ func (e *Endpoint) GetCIDRPrefixLengths() (s6, s4 []int) {
+					Ingress: make(policy.DirectionalVisibilityPolicy),
+					Egress:  make(policy.DirectionalVisibilityPolicy),
+				}
+				return
```
This seems odd. Why are we enforcing a policy recalculation in like 754 if `err` is `nil` but not for this case?

ianvernon replied:
Good catch, fixed.

aanm reviewed (commented):
It would have been easier to review if there were more commits across the PR

Commit:
policy: add `ProxyPolicy` interface

This will be used to allow for multiple types to be passed into functions
related to proxy configuration in upcoming commits.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: add `VisibilityPolicy` type

This type encodes information about what traffic should be redirected to a
given L7 proxy for both traffic directions (ingress and egress).

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: add `VisibilityPolicy` field and plumbing logic

When a `VisibilityPolicy` exists for an endpoint, and policy enforcement is
disabled for the traffic direction of the `VisibilityPolicy`, an entry is added
to the `PolicyMap` for the endpoint which will redirect the traffic to the L7
proxy. If policy enforcement is enabled for a given traffic direction, then
`VisibilityPolicy` currently is ignored.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
proxy: change functions to take `policy.ProxyPolicy` parameter

This will allow for the proxy functions to take in an `L4Filter` or a
`VisibilityPolicy` when redirects are being generated.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
annotation: add annotation for proxy visibility

Future commits will check if a pod has been annotated with this annotation
upon receiving a pod update event from the Kubernetes watcher. If so, then the
endpoint corresponding to the pod will have its `VisibilityPolicy` updated
accordingly.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
daemon: get pod annotations along with labels from K8s

When we receive the labels for a pod, we also should receive its annotations so
that we can update the `VisibilityPolicy` for the endpoint corresponding to the
pod.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
k8s: check if proxy visibility annotation has changed upon pod update

If so, update the `VisibilityPolicy` accordingly for the Endpoint. Plumb this
logic into the K8s watcher for the daemon.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test: add e2e test for proxy-visibility

Add a basic Ginkgo test to the K8s test suite which tests the following:

* With no proxy visibility annotation, no traffic is redirected to the proxy.
* Adding the proxy visbility annotation redirects traffic to the proxy.
* Removing the proxy visibility annotation removes the redirect, and traffic is
  not redirected to the proxy.

Signed-off by: Ian Vernon <ian@cilium.io>

ianvernon commented on pkg/proxy/redirect.go:
```diff
@@ -15,9 +15,10 @@
-	"github.com/cilium/cilium/pkg/revert"
 	"time"
 
+	"github.com/cilium/cilium/pkg/revert"
```
TODO: fix import order.

joestringer replied:
^^^

(If you set up your editor to run `go fmt` on save, then you can just put it in the same block as the rest of the cilium bits and it will resolve this for you..)

ianvernon commented on test/k8sT/Policies.go:
```diff
@@ -719,6 +726,119 @@ var _ = Describe("K8sPolicyTest", func() {
+				app1Pod   string
+				ciliumPod string
+				nodeName  string
+				//app1EpID        int64
```
Remove this.

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -242,31 +251,152 @@ func (e *Endpoint) addNewRedirectsFromMap(m policy.L4PolicyMap, desiredRedirects
+	if !policyEnabled {
+		for _, visMeta := range visPolicy {
+			// Create a redirect for every entry in the visibility policy.
+			if !e.hasSidecarProxy || visMeta.Parser != policy.ParserTypeHTTP {
```
Can you flip this condition and do a continue?
```golang
			if e.hasSidecarProxy && visMeta.Parser == policy.ParserTypeHTTP {
				continue
			}
```

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -242,31 +251,152 @@ func (e *Endpoint) addNewRedirectsFromMap(m policy.L4PolicyMap, desiredRedirects
+		policyEnabled = e.desiredPolicy.EgressPolicyEnabled
+	}
+
+	if !policyEnabled {
```
If `policyEnable` is true we should not even continue at all, including doing a `revertStack.Push` in line 335

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -242,31 +251,152 @@ func (e *Endpoint) addNewRedirectsFromMap(m policy.L4PolicyMap, desiredRedirects
 
-	var ff revert.FinalizeFunc
-	var rf revert.RevertFunc
+	// Ingress redirects.
```
can't we have a for loop here?

```
for ingress := range  []bool{true, false} {
// code from lines 370 to 384
}
```

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -242,31 +251,152 @@ func (e *Endpoint) addNewRedirectsFromMap(m policy.L4PolicyMap, desiredRedirects
+		for _, visMeta := range visPolicy {
+			// Create a redirect for every entry in the visibility policy.
+			if !e.hasSidecarProxy || visMeta.Parser != policy.ParserTypeHTTP {
+				var (
```
A lot of this code seems similar to the code present in `addNewRedirectsFromMap`. Can't we refactor it?

ianvernon replied:
The first iteration of this code handles the case in which policy is not-enabled. Future work will work on adding redirects based off of the state of the L4Policy which is generated for the Endpoint, which I believe will cause divergence between `addNewVisibilityRedirects` and `addNewRedirectsFromMap`. So, I would prefer keeping them separate for now.

aanm commented on pkg/endpoint/redirect_test.go:
```diff
@@ -0,0 +1,256 @@
+	_, err, _, _ = ep.addNewRedirects(ep.desiredPolicy.L4Policy, cmp)
+	c.Assert(err, check.IsNil)
+	v, ok := ep.desiredPolicy.PolicyMapState[policy.Key{
+		Identity:         0,
```
Why is identity zero here?

ianvernon replied:
Identity of zero is an allow-all in our datapath. Proxy redirection occurs for a given L4-port / protocol - this means it allows-all at L3.

joestringer replied:
We should probably add a constant for this somewhere ;)

aanm commented on pkg/policy/l4.go:
```diff
@@ -132,6 +132,31 @@ type L4Filter struct {
 }
 
+// CopyL7RulesPerEndpoint creates a copy of the L7RulesPerEp of the L4Filter.
+func (l4 *L4Filter) CopyL7RulesPerEndpoint() L7DataMap {
```
just a warning this is a not a deepcopy.

ianvernon replied:
👍  - the old behavior did exactly this, just inline in `redirect.updateRules`:

```
@@ -73,12 +73,9 @@ func newRedirect(localEndpoint logger.EndpointUpdater, listener *ProxyPort, dstP
 // updateRules updates the rules of the redirect, Redirect.mutex must be held
 // 'implementation' is not initialized when this is called the first time.
 // TODO: Replace this with RedirectImplementation UpdateRules method!
-func (r *Redirect) updateRules(l4 *policy.L4Filter) revert.RevertFunc {
+func (r *Redirect) updateRules(p policy.ProxyPolicy) revert.RevertFunc {
        oldRules := r.rules
-       r.rules = make(policy.L7DataMap, len(l4.L7RulesPerEp))
-       for key, val := range l4.L7RulesPerEp {
-               r.rules[key] = val
-       }
+       r.rules = p.CopyL7RulesPerEndpoint()
        return func() error {
                r.mutex.Lock()
                r.rules = oldRules
```

so there is no functional change here.

aanm commented on pkg/policy/l4.go:
```diff
@@ -132,6 +132,31 @@ type L4Filter struct {
 
+// CopyL7RulesPerEndpoint creates a copy of the L7RulesPerEp of the L4Filter.
+func (l4 *L4Filter) CopyL7RulesPerEndpoint() L7DataMap {
+	m := make(L7DataMap, len(l4.L7RulesPerEp))
```
This could be a method of `L7DataMap` instead

aanm commented on pkg/policy/visibility.go:
```diff
@@ -0,0 +1,181 @@
+	"github.com/cilium/cilium/pkg/u8proto"
+)
+
+var annotationRegex = regexp.MustCompile("^((<(Ingress|Egress)/([0-9]{1,6})/(TCP|UDP|ANY)/([A-Za-z]{3,32})>,)+)$")
```
Shouldn't the regex be 
```
`^((<(Ingress|Egress)/([0-9]{1,6})/(TCP|UDP|ANY)/([A-Za-z]{3,32})>)(,(<(Ingress|Egress)/([0-9]{1,6})/(TCP|UDP|ANY)/([A-Za-z]{3,32})>))*)$`
```

ianvernon replied:
I tried to get this, but didn't think of constructing the regex this way; your way is much better. Will update accordingly.

aanm commented on pkg/policy/visibility.go:
```diff
@@ -0,0 +1,181 @@
+//   different L7 protocols for the same L4 port / protocol / traffic direction.
+func NewVisibilityPolicy(anno string) (*VisibilityPolicy, error) {
+	// Add a trailing comma so we can match the regex for now.
+	if anno[len(anno)-1] != ',' {
```
No need for this with the proposed regex

aanm commented on pkg/policy/visibility.go:
```diff
@@ -0,0 +1,181 @@
+	anSplit := strings.Split(anno, ",")
+	for i := range anSplit {
+		// Avoid empty string
+		if len(anSplit[i]) == 0 {
```
How can this happen if the regex does not allow it?

ianvernon replied:
I added this for some reason while developing this, but over time it turned into dead code. Removed.

aanm commented on pkg/policy/visibility.go:
```diff
@@ -0,0 +1,181 @@
+			return nil, err
+		}
+		// <Ingress|Egress --> Ingress|Egress
+		// Don't need to validate the content itself, regex already did that.
```
If this is true, why do we need to do validation on line 75 and line 85?

ianvernon replied:
I did so out of an abundance of caution - it shouldn't be necessary, but given that I am directly dereferencing specific indices in the result of `strings.Split` I figured it would be good to make sure that that would be the case instead of causing a panic.

aanm commented on pkg/policy/visibility.go:
```diff
@@ -0,0 +1,181 @@
+		Egress:  make(DirectionalVisibilityPolicy),
+	}
+
+	anSplit := strings.Split(anno, ",")
```
Just a small note that the work being done on lines 69-116 could be achieved with regex groups.

ianvernon replied:
I've added a TODO here for now; would you prefer I take this up as part of this PR?

aanm replied:
Can be done later

aanm commented on pkg/policy/visibility_test.go:
```diff
@@ -0,0 +1,66 @@
+// +build !privileged_tests
```
header missing

Commit:
endpoint: remove useless nil check

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: reduce duplicate code when generating ingress and egress redirects

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: reduce amount of code contained in if-statement

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: only push to revert stack if policy not enabled

If policy is enabled, pushing onto the revert stack in this case is a no-op
since none of the structures which are iterated over contain any data.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: factor out L7DataMap shallow copying into function

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: improve regex for proxy-visibility annotation

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: add copyright header to `visibility_test.go`

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: add TODO for using regular expression groups when parsing annotation

Signed-off by: Ian Vernon <ian@cilium.io>

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -366,41 +366,29 @@ func (e *Endpoint) addNewRedirects(m *policy.L4Policy, proxyWaitGroup *completio
-		if err != nil {
-			return desiredRedirects, fmt.Errorf("unable to allocate ingress redirects: %s", err), nil, nil
+	for _, ingress := range []bool{true, false} {
+		dirLogStr := "ingress"
```
really small nil, you could pass a map into the for loop

```golang
	for dirLogStr, ingress := range map[string]bool{"ingress": true, "egress": false} {
```

ianvernon replied:
Done!

aanm reviewed (commented):
2 small nits.

I thought we would be better off by having commits created after a person have reviewed a PR but it was more painful until I decided to review the PR backwards (starting from the last commit up to the last commit reviewed from the last review)

I assume you will be squashing the post-review commits together with the pre-review commits so that we have a logical change per commit?

I don't think I have anything blocking at this point.

Commit:
endpoint: map traffic direction boolean to corresponding string for logging

This is more concise; no functional changes here.

Signed-off by: Ian Vernon <ian@cilium.io>

jrajahalme commented on pkg/policy/visibility.go:
```diff
@@ -0,0 +1,181 @@
+		// Don't need to validate the content itself, regex already did that.
+		direction := proxyAnnoSplit[0][1:]
+		port := proxyAnnoSplit[1]
+		portInt, err := strconv.Atoi(port)
```
`portInt` the values <=0 and >65535 should be explicitly rejected.

jrajahalme commented on pkg/endpoint/bpf.go:
```diff
@@ -242,31 +251,152 @@ func (e *Endpoint) addNewRedirectsFromMap(m policy.L4PolicyMap, desiredRedirects
+	err, ff, rf = e.addVisibilityRedirects(true, desiredRedirects, proxyWaitGroup)
 	if err != nil {
-		return desiredRedirects, fmt.Errorf("unable to allocate ingress redirects: %s", err), nil, nil
+		return desiredRedirects, fmt.Errorf("unable to allocate ingress visibility redirects: %s", err), nil, nil
```
Should revert the reverts stack before returning.

jrajahalme commented on pkg/endpoint/bpf.go:
```diff
@@ -242,31 +251,152 @@ func (e *Endpoint) addNewRedirectsFromMap(m policy.L4PolicyMap, desiredRedirects
 	if err != nil {
-		revertStack.Revert() // Ignore errors while reverting. This is best-effort.
-		return desiredRedirects, fmt.Errorf("unable to allocate egress redirects: %s", err), nil, nil
+		return desiredRedirects, fmt.Errorf("unable to allocate ingress visibility redirects: %s", err), nil, nil
```
should revert the revert stack before returning with an error. This could also be done with a defer.

jrajahalme commented on daemon/k8s_watcher.go:
```diff
@@ -1567,22 +1578,51 @@ func (d *Daemon) updateK8sPodV1(oldK8sPod, newK8sPod *types.Pod) error {
-	newLabels := labels.Map2Labels(newPodLabels, labels.LabelSourceK8s)
+	switch {
+	case annotationsChanged && labelsChanged:
+		// Update annotations and let identity update handle regeneration.
```
identity may not have actually changed, so ModifyIdentityLabels() may not regenerate in all cases.

ianvernon replied:
But the labels themselves have changed - we calculated that above. Is there a possibility the labels may not have changed in spite of that?

ianvernon replied:
Discussed offline - added an explicit call to regenerate in this code path in the last commit on this PR.

jrajahalme commented on pkg/k8s/factory_functions.go:
```diff
@@ -194,12 +194,29 @@ func EqualV2CNP(cnp1, cnp2 *types.SlimCNP) bool {
+// relevantAnnotations is equal in anno1 and anno2.
+func AnnotationsEqual(relevantAnnotations []string, anno1, anno2 map[string]string) bool {
+	for i := range relevantAnnotations {
+		an := relevantAnnotations[i]
```
would be simpler to `for _, an := range relevantAnnotations {`

jrajahalme commented on test/k8sT/Policies.go:
```diff
@@ -719,6 +726,119 @@ var _ = Describe("K8sPolicyTest", func() {
+				By("Checking that aforementioned traffic was%sredirected to the proxy", not)
+				reStr := fmt.Sprintf("verdict Forwarded GET http://%s/public", app1PodIP)
+				re := regexp.MustCompile(reStr)
+				out := re.Find(monitorOutput)
```
Could use the ExpectContains() helper instead if the monitor output would not be redirected to a file (like in runtime/fqdn.go)

ianvernon replied:
I personally like redirecting to file because that way we get the monitor files in the logs for free that way.

jrajahalme commented on pkg/endpoint/bpf.go:
```diff
@@ -366,41 +366,29 @@ func (e *Endpoint) addNewRedirects(m *policy.L4Policy, proxyWaitGroup *completio
+		if !ingress {
+			dirLogStr = "egress"
+		}
+		// Ingress redirects.
```
Comment is now wrong

jrajahalme commented on pkg/endpoint/bpf.go:
```diff
@@ -366,41 +366,29 @@ func (e *Endpoint) addNewRedirects(m *policy.L4Policy, proxyWaitGroup *completio
+		if m != nil && m.HasRedirect() {
+			err, ff, rf = e.addNewRedirectsFromMap(ingress, desiredRedirects, proxyWaitGroup)
+			if err != nil {
+				return desiredRedirects, fmt.Errorf("unable to allocate %s redirects: %s", dirLogStr, err), nil, nil
```
must revert the revert stack if returning with an error.

ianvernon replied:
I believe this function is good to go now w.r.t. reverting. Good catch!

jrajahalme replied:
Like the defer 👍

joestringer replied:
Personally I think the defer is a bit harder to read/understand than just explicitly calling it at each error return spot. But that said I think the implementation is correct in its current form, so I won't press on changing this.

jrajahalme commented on pkg/endpoint/bpf.go:
```diff
@@ -366,41 +366,29 @@ func (e *Endpoint) addNewRedirects(m *policy.L4Policy, proxyWaitGroup *completio
 		if err != nil {
-			revertStack.Revert() // Ignore errors while reverting. This is best-effort.
-			return desiredRedirects, fmt.Errorf("unable to allocate egress redirects: %s", err), nil, nil
+			return desiredRedirects, fmt.Errorf("unable to allocate %s visibility redirects: %s", dirLogStr, err), nil, nil
```
must read evert the revert stack if returning with an error.

ianvernon replied:
I believe this function is good to go now w.r.t. reverting. Good catch!

jrajahalme reviewed (changes requested):
Mostly nits but the reverts need to be done on error returns as indicated in the comments.

Commit:
policy: restrict ports in annotation to be in range (0,63356]

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: revert partially successful redirect allocation if subsequent allocations fail

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
k8s: simplify for loop iteration in `AnnotationsEqual`

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: remove outdated comment

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
k8s: explicitly regenerate when pod annotations updated

In the case where we receive a pod update which contains both label updates and
annotation updates, we are not guaranteed to actually regenerate the endpoint if
the labels have changed because the logic internal to processing the labels
update on the endpoint side may skip regenerating. Consequently, always
regenerate the Endpoint explicitly for the annotation update as well. This may
add an unnecessary regeneration for the Endpoint, but it guarantees we will
correctly realize the state prescribed via Kubernetes.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
k8s: change handling of labels + annotation updates for pods

Always try to update the labels first. If updating the labels fails, return the
error immediately. Then, if no error has occurred or no pod label updates are
needed, update the visibility policy + regenerate the Endpoint.

Signed-off by: Ian Vernon <ian@cilium.io>

aanm reviewed (approved):
I think overall this looks good. I believe this needs a couple cleanups from @ianvernon before being merged.

joestringer commented on daemon/k8s_watcher.go:
```diff
@@ -1566,22 +1577,51 @@ func (d *Daemon) updateK8sPodV1(oldK8sPod, newK8sPod *types.Pod) error {
+	return nil
+}
+
+func realizePodAnnotationUpdate(podEP *endpoint.Endpoint) {
```
nit: Is there something that necessitates putting this logic (regeneration trigger) in `daemon/k8s_watcher.go` rather than under endpoint? The equivalent labels handling is all done in `pkg/endpoint/endpoint.go`, including the regeneration.

(Not consequential; feel free to ignore)

ianvernon replied:
I initially had two callsites for this from the `daemon/k8s_watcher.go` code, but this ended up being the only one.

joestringer commented on daemon/state.go:
```diff
@@ -70,10 +70,13 @@ func (d *Daemon) validateEndpoint(ep *endpoint.Endpoint) (valid bool, err error)
 			return false, fmt.Errorf("kubernetes pod not found")
 		}
+		if err == nil {
+			ep.UpdateVisibilityPolicy(p.Annotations)
```
The pod update case in the k8s watcher triggers regeneration, but this path doesn't. Is this because we will regenerate the endpoint using the latest information from k8s anyway, so there's no use queuing up another regeneration?

Organizational nit, this function is named `validateEndpoint` and the comment suggests that it is merely confirming that the endpoint structure as restored is the same as in k8s, but this change actually means it has side-effects; it will modify/update the visibility policy. I think it'd be nice to at least mention this in the comment above the function.

ianvernon replied:
The restore logic handles regeneration in quite a special way; regeneration is triggered here: https://github.com/cilium/cilium/blob/75c77f1c1c273091d26af31a1485fc0090a181ec/daemon/state.go#L200 .

Agreed re: code organization. Will update the comment. I wanted to piggy-back on top of existing logic since we already are querying the API here.

joestringer replied:
OK, I thought it was to do with the API query here. Might also be useful to leave a comment in the code pointing out how the regeneration works. Up to you.

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -145,14 +145,24 @@ func (e *Endpoint) writeHeaderfile(prefix string) error {
 // problem that occurred while adding an l7 redirect for the specified policy.
 // Must be called with endpoint.Mutex held.
-func (e *Endpoint) addNewRedirectsFromMap(m policy.L4PolicyMap, desiredRedirects map[string]bool, proxyWaitGroup *completion.WaitGroup) (error, revert.FinalizeFunc, revert.RevertFunc) {
+func (e *Endpoint) addNewRedirectsFromMap(ingress bool, desiredRedirects map[string]bool, proxyWaitGroup *completion.WaitGroup) (error, revert.FinalizeFunc, revert.RevertFunc) {
```
nit, if this is switching to `ingress bool` then the previous name doesn't make sense any more:

```suggestion
func (e *Endpoint) addNewRedirectsFromDesiredPolicy(ingress bool, desiredRedirects map[string]bool, proxyWaitGroup *completion.WaitGroup) (error, revert.FinalizeFunc, revert.RevertFunc) {
```

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -242,34 +251,153 @@ func (e *Endpoint) addNewRedirectsFromMap(m policy.L4PolicyMap, desiredRedirects
+
+	for _, visMeta := range visPolicy {
+		// Create a redirect for every entry in the visibility policy.
+		if e.hasSidecarProxy && visMeta.Parser == policy.ParserTypeHTTP {
```
This check is logically correct, the semantics of `hasSidecarProxy` is that Envoy is running inside the endpoint, and we will configure that instance, so all traffic will go via the proxy anyway therefore we can skip the policy.

What I'm wondering / concerned about is how do we find this logic if our assumptions change? For instance, if we end up handling Kafka via Envoy/sidecar proxy, which seems plausible in the mid-to-long term.

Simplest answer to this is "we'll deal with it when we change that", though maybe it'd be slightly easier to find if we had a clear factored out function that did something like:

```
func (e *Endpoint) SkipProxyRedirect(parser policy.L7ParserType) bool {
    return e.hasSidecarProxy && parser == policy.ParserTypeHTTP
}
```

(Feel free to ignore if it seems OTT)

ianvernon replied:
Created https://github.com/cilium/cilium/issues/9356 to track this.

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -242,34 +251,153 @@ func (e *Endpoint) addNewRedirectsFromMap(m policy.L4PolicyMap, desiredRedirects
 	return nil, finalizeList.Finalize, revertStack.Revert
 }
 
+func (e *Endpoint) addVisibilityRedirects(ingress bool, desiredRedirects map[string]bool, proxyWaitGroup *completion.WaitGroup) (error, revert.FinalizeFunc, revert.RevertFunc) {
```
Did you consider refactoring this function / `addNewRedirectsFromMap()` to have the same code path?

I'm concerned that by having two very similar functions for adding redirects, we increase maintenance burden and likelihood of introducing bugs (or failing to fix in one of the cases) when future changes are made in this area of the code. Also, if the codepaths interact or depend on the implementation of the other code path, then it's harder to distinguish the subtleties when the details are in two places.

ianvernon replied:
I am anticipating the code being different once I implement the case where we add visibility on top of existing policy rules that apply to an endpoint. If the code paths end up being collapsible, I will collapse them together. The first iteration is meant to unblock folks who are dependent upon this functionality being added soon.

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -242,34 +251,153 @@ func (e *Endpoint) addNewRedirectsFromMap(m policy.L4PolicyMap, desiredRedirects
-		revertStack.Revert() // Ignore errors while reverting. This is best-effort.
-		return desiredRedirects, fmt.Errorf("unable to allocate egress redirects: %s", err), nil, nil
+	for dirLogStr, ingress := range map[string]bool{"ingress": true, "egress": false} {
+		if m != nil && m.HasRedirect() {
```
I don't think that this conditional check is necessary.

* The caller already checks that `m != nil`, otherwise it shouldn't call this function at all.
* `m.HasRedirect()` runs through every filter in both the ingress AND egress directions to check if `filter.isRedirect()`, but inside of the function `addNewRedirectsFromMap()` called just one line down inside this conditional block, it *also* iterates every filter, but specifically in the direction that is specified; it checks if the filter `IsRedirect()` as well. So this check is actually iterating over all filters (including irrelevant ones for the direction being handled) unnecessarily.

joestringer commented on pkg/endpoint/policy.go:
```diff
@@ -709,3 +710,48 @@ func (e *Endpoint) GetCIDRPrefixLengths() (s6, s4 []int) {
+		// is needed to update endpoint policy maps with the policy map state
+		// generated from the visibility policy. This can, and should be more
+		// elegant in the future.
+		e.forcePolicyComputation()
```
Should this only be done at the end of the function (ie in the success case)? If `NewVisibilityPolicy()` fails, this will not only fail to apply visibility policy, but also unnecessarily force computation.

ianvernon replied:
Note that below we disable visibility policy if an invalid visibility policy is imported. So, we are always changing the visibility policy in either success or failure cases. In order to realize this state in either case we have to force computation of policy.

joestringer replied:
OK, thanks! I didn't glean this from the comment here, if you think it's useful to leave this context in the code then maybe mention that as well.

joestringer commented on pkg/endpoint/policy.go:
```diff
@@ -709,3 +710,48 @@ func (e *Endpoint) GetCIDRPrefixLengths() (s6, s4 []int) {
+// Endpoint will be empty, and will have no effect. If the proxy visibility
+// annotation cannot be parsed, an empty visibility policy is assigned to the
+// Endpoint.
+func (e *Endpoint) UpdateVisibilityPolicy(anno map[string]string) {
```
nit: This is half-way between being explicit about policy and being generic to handle any annotations. Do you expect that the endpoint package should be responsible for understanding specific annotations going forward?

A couple of suggestions:
* `UpdateAnnotations(anno map[string]string)` would be more general and establish that `pkg/endpoint` should know about annotation keys; alternatively,
* `UpdateVisibilityPolicy(annotation string)` would be more specific (callers would call `UpdateVisibilityPolicy(annotations[annotation.ProxyVisibility])`), which would allow the endpoint package to be oblivious of the annotation key value and not depend on `pkg/annotation`.

ianvernon replied:
I'll implement it as you like. I personally err on the side of having more generic functionality in the parameter list for functions to avoid having to refactor all callsites in the future; if we add another annotation we care about, we get it for free in the current form. But, I don't really have a preference so I will defer to you here.

joestringer replied:
If you want it to be generic to reduce need for refactoring in future, then `UpdateAnnotations()` would make more sense. If we left it as-is, then in future you'd need to at least rename the function anyway because most likely any new annotations would have their own semantics and potentially wouldn't even be related to visibility policy at all.

joestringer commented on pkg/endpoint/proxy.go:
```diff
@@ -69,16 +69,23 @@ func (e *Endpoint) removeNetworkPolicy() {
 	e.proxy.RemoveNetworkPolicy(e)
 }
 
+// FakeEndpointProxy is a stub proxy used for testing.
```
Apparently this code already existed, but consider this comment a brief reminder that `pkg/testutils` exists to host common fake test-related packages like this.

joestringer commented on pkg/endpoint/redirect_test.go:
```diff
@@ -0,0 +1,256 @@
+	}
+
+	ep.UpdateVisibilityPolicy(secondAnno)
+	d, err, _, _ := ep.addNewRedirects(ep.desiredPolicy.L4Policy, cmp)
```
Doesn't `secondAnno` conflict with `firstAnno`? Should it return an error?

`firstAnno`: `"<Ingress/80/TCP/HTTP>"`
`secondAnno`: `"<Ingress/80/TCP/Kafka>"`

ianvernon replied:
The value provided to `UpdateVisibilityPolicy` describes the desired state; this means that if we call `UpdateVisibilityPolicy(firstAnno)`, and `UpdateVisibilityPolicy(secondAnno)`, the redirect from firstAnno will be removed, as secondAnno describes the desired state. Conflicts can only occur if in the entire annotation string there is a conflict, e.g. `<Ingress/80/TCP/HTTP>,<Ingress/80/TCP/Kafka>`

joestringer replied:
I read this the wrong way, makes sense. Thanks for explaining.

joestringer commented on pkg/endpoint/redirect_test.go:
```diff
@@ -0,0 +1,256 @@
+		TrafficDirection: trafficdirection.Ingress.Uint8(),
+	}]
+	c.Assert(ok, check.Equals, true)
+	c.Assert(v.ProxyPort, check.Equals, kafkaPort)
```
What's the difference between the hunks on lines 211-218 and 220-227?

ianvernon replied:
Ingress vs. Egress

joestringer replied:
No, egress is handled on lines 229-236.

ianvernon replied:
You're right, I don't know how this got there... removing it.

joestringer commented on pkg/k8s/labels.go:
```diff
@@ -95,7 +95,7 @@ func isInjectedWithIstioSidecarProxy(scopedLog *logrus.Entry, pod *corev1.Pod) b
 
 // GetPodLabels returns the labels of a pod
-func GetPodLabels(namespace, podName string) (map[string]string, error) {
+func GetPodLabels(namespace, podName string) (lbls map[string]string, retAnno map[string]string, retErr error) {
```
This is no longer just getting pod labels, should it be renamed?

```suggestion
func GetPodMetadata(namespace, podName string) (lbls map[string]string, retAnno map[string]string, retErr error) {
```

joestringer commented on pkg/policy/visibility.go:
```diff
@@ -0,0 +1,179 @@
+	"github.com/cilium/cilium/pkg/u8proto"
+)
+
+var annotationRegex = regexp.MustCompile(`^((<(Ingress|Egress)/([0-9]{1,6})/(TCP|UDP|ANY)/([A-Za-z]{3,32})>)(,(<(Ingress|Egress)/([0-9]{1,5})/(TCP|UDP|ANY)/([A-Za-z]{3,32})>))*)$`)
```
nit: the port here is `[0-9]{1,6}`, but should it be `[1-9][0-9]{0,5}`? All-zeros port (or zero-prefixed port) is, I believe, nonsensical.

Also, not sure how the limits work here, but is this `6` inclusive or exclusive? I don't see a test with 6-digit input.

Any chance that the second half where it copies the same expression to allow a list of rediercts, could instead be represented using the `{1,X}` notation, where we'd figure out what the maximum is given the current character limits? I guess there's still the comma in between that makes it a bit trickier, hmm.. But I guess we could at least declare the regex for the annotation once, then combine strings here so we can ensure that they're the same (in case of any future changes). Example:

```
var (
    singleAnnotationRegex = '(<(Ingress|Egress)/([0-9]{1,6})/(TCP|UDP|ANY)/([A-Za-z]{3,32})>)'
    annotationRegex = regexp.MustCompile(fmt.Sprintf('^(%s)(,%s)*)$', singleAnnotationRegex, singleAnnotationRegex))
)
```

joestringer commented on pkg/policy/visibility.go:
```diff
@@ -0,0 +1,179 @@
+		direction := proxyAnnoSplit[0][1:]
+		port := proxyAnnoSplit[1]
+
+		portInt, err := strconv.ParseUint(port, 0, 16)
```
Shouldn't this be base 10?

```suggestion
		portInt, err := strconv.ParseUint(port, 10, 16)
```

joestringer commented on pkg/policy/visibility.go:
```diff
@@ -0,0 +1,179 @@
+		}
+
+		if portInt == 0 {
+			return nil, fmt.Errorf("port cannot be 0")
```
I think we could exclude this case using regex. Unit tests should confirm the rest.

joestringer commented on test/k8sT/Policies.go:
```diff
@@ -719,6 +726,119 @@ var _ = Describe("K8sPolicyTest", func() {
+				monitorStop := kubectl.MonitorStart(helpers.KubeSystemNamespace, ciliumPod, monitorFile)
+
+				// Let the monitor get started since it is started in the background.
+				time.Sleep(2 * time.Second)
```
I'm surprised that these kinds of sleeps are necessary for getting reliable monitor output in tests, I hope that we can drastically improve this. I'm not sure why the same was not necessary in the monitor aggregation tests, is that test just relying on some race condition?

ianvernon replied:
Yeah, this one took me a while to figure out. We are launching the monitor in the background, so there could be some lag between the launching of the monitor which causes it to not capture the flow for the `curl` call below.

joestringer replied:
https://github.com/cilium/cilium/issues/9357

joestringer commented on test/k8sT/Policies.go:
```diff
@@ -719,6 +726,119 @@ var _ = Describe("K8sPolicyTest", func() {
+					namespaceForTest, appPods[helpers.App2],
+					curlCmd)
+				// Give time for the monitor to be notified of the proxy flow.
+				time.Sleep(2 * time.Second)
```
Ah, this is the part that is missing in monitor aggregation which I believe caused the flake #9330 .

joestringer commented on test/k8sT/Policies.go:
```diff
@@ -719,6 +726,119 @@ var _ = Describe("K8sPolicyTest", func() {
+				Expect(kubectl.CiliumEndpointWaitReady()).To(BeNil())
+
+				checkProxyRedirection(app1PodIP, false)
+
```
I recall you mentioned that you've tested the following:

1. No enforcement policy
1. Apply visibility policy
1. Apply enforcement policy that drops the traffic affected by visibility
1. Now the redirect no longer occurs

Are you planning to add this to the regression suite as well?

ianvernon replied:
Added this test.

joestringer reviewed (changes requested):
Lots of nits, the main things I'd focus on are:

* Remove the extra unnecessary iteration of filters
* Double-check regex bits, make them a bit easier to read
* Feedback on the test files

But LGTM otherwise.

Commit:
daemon: add comment explaining context around `UpdateVisibilityPolicy`

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: rename `addNewRedirectsFromMap`

The `L4PolicyMap` is no longer passed in as a parameter, and we are accessing
the `DesiredPolicy` when calling this function; rename accordingly.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: remove check around `addNewRedirectsFromDesiredPolicy`

The caller checks whether the map is nil, so we do not need to check it here. We
also do not need to iterate over all filters with the call to
`L4PolicyMap.HasRedirect()`; we are already checking whether each `L4Filter` has
a redirect further down the call stack.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: refactor `UpdateVisibilityPolicy` to take in a string

This removes the need to import `pkg/annotation` from `pkg/endpoint`.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
k8s: rename `GetPodLabels` to `GetPodMetadata`

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: factor out duplicate regex string into variable

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: restrict values for port in visibility annotation

Do not allow leading zeroes. Add a unit test for this as well as
a test which checks that we do not allow > 5 digits.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: restrict port to be base 10 in visibility annotation

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: remove unneeded zero-port check

The annotation regex takes care of this now.

Signed-off by: Ian Vernon <ian@cilium.io>

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -242,34 +251,153 @@ func (e *Endpoint) addNewRedirectsFromMap(m policy.L4PolicyMap, desiredRedirects
-		return desiredRedirects, fmt.Errorf("unable to allocate egress redirects: %s", err), nil, nil
+	for dirLogStr, ingress := range map[string]bool{"ingress": true, "egress": false} {
+		if m != nil && m.HasRedirect() {
+			err, ff, rf = e.addNewRedirectsFromMap(ingress, desiredRedirects, proxyWaitGroup)
```
Note: This change in passing down the policy means that `addNewRedirects()` effectively ignores the first parameter, the policy. Either `addNewRedirects()` should drop the policy argument, or this code should actually pass down some part of the policy that is specified in the first parameter to `addNewRedirects()`. (I'm guessing the former)

ianvernon replied:
Fixed.

Commit:
test: add test which tests visibility / policy interactions

Add a test which ensures the following behavior occurs:
* Proxy visibility is removed when policy is imported which selects an endpoint
with a visibility policy.
* Proxy visibility is reinstated when policy no longer selects an endpoint with
a visibility policy.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: remove unused parameter from `addNewRedirects`

Signed-off by: Ian Vernon <ian@cilium.io>
